[Core] Fix grpc server is started check

kfstorm opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

If the grpc server port is specified (`port_` is not 0), `port_` will not become 0 if the grpc server failed to start. According to [grpc doc](https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_builder.html#a3ad1429e6575de79f8f17dad383fbc04), we should check if the return value of `BuildAndStart` is nullptr.

Raylet will report a more meaningful error message with this PR rather than crashing in `CreateCall` like #8254. 

Note: This is not the fix of #9615. 

## Related issue number

#9615

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -59,19 +59,27 @@ void GrpcServer::Run() {
-      << "Port " << specified_port
-      << " specified by caller already in use. Try passing node_manager_port=... into "
-         "ray.init() to pick a specific port";
+  std::ostringstream ss;
```
move it to inside if (!server) block?

rkooo567 commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -59,19 +59,27 @@ void GrpcServer::Run() {
+         << "and maybe it's already in use. Try passing node_manager_port=... into "
+            "ray.init() to pick a specific port";
+    } else {
+      ss << " No specified port.";
```
This error seems to be pretty ambiguous. Can you make it a bit clearer what's the problem here?

kfstorm replied:
Maybe I'll just remove the port part message because grpc already outputs the reason?

kfstorm commented:
@rkooo567 I've updated code and added a test case.

Note that I deleted `grpc_server_num_retries` because I get below crash log when retry is enabled.

```
E0218 17:11:01.061957000 4550471168 server_cc.cc:1083]                 assertion failed: service->server_ == nullptr && "Can only register an asynchronous service against one server."
[2021-02-18 17:11:01,062 E 8027 1757351] logging.cc:435: *** Aborted at 1613639461 (unix time) try "date -d @1613639461" if you are using GNU date ***
[2021-02-18 17:11:01,062 E 8027 1757351] logging.cc:435: PC: @                0x0 (unknown)
[2021-02-18 17:11:01,062 E 8027 1757351] logging.cc:435: *** SIGABRT (@0x7fff2052b462) received by PID 8027 (TID 0x10f3aae00) stack trace: ***
[2021-02-18 17:11:01,062 E 8027 1757351] logging.cc:435:     @     0x7fff2059dd7d _sigtramp
[2021-02-18 17:11:01,062 E 8027 1757351] logging.cc:435:     @     0x7fd0970beda0 (unknown)
[2021-02-18 17:11:01,062 E 8027 1757351] logging.cc:435:     @     0x7fff204ac720 abort
[2021-02-18 17:11:01,063 E 8027 1757351] logging.cc:435:     @        0x10a1d2a0d grpc_impl::Server::RegisterService()
[2021-02-18 17:11:01,064 E 8027 1757351] logging.cc:435:     @        0x109df39f9 grpc_impl::Server::RegisterService()
[2021-02-18 17:11:01,064 E 8027 1757351] logging.cc:435:     @        0x109def8f3 grpc_impl::ServerBuilder::BuildAndStart()
[2021-02-18 17:11:01,064 E 8027 1757351] logging.cc:435:     @        0x109c7c375 ray::rpc::GrpcServer::Run()
[2021-02-18 17:11:01,064 E 8027 1757351] logging.cc:435:     @        0x10986fc8f ray::raylet::NodeManager::NodeManager()
...
```

ijrsvt reviewed (approved):
This looks good to me! Thanks for fixing this issue with surfacing failures here!

rkooo567 commented on python/ray/cluster_utils.py:
```diff
@@ -100,20 +101,21 @@ def add_node(self, **node_args):
-        # tests we assume that the driver is connected to the first node that
-        # is added.
-        self._wait_for_node(node)
+        if wait:
```
This will slow down CI a lot I guess. Is it to avoid the port conflicting?

rkooo567 replied:
Ah, if it is for testing, can we just make the default False?

rkooo567 replied:
(so that we can keep the previous behavior)

kfstorm replied:
The previous behavior is to wait always, so I didn't change the default behavior.

rkooo567 commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -30,48 +30,35 @@ GrpcServer::GrpcServer(std::string name, const uint32_t port, int num_threads)
-      << "Port " << specified_port
-      << " specified by caller already in use. Try passing node_manager_port=... into "
-         "ray.init() to pick a specific port";
+  RAY_CHECK(server_) << "Failed to start the grpc server.";
```
Can we add which port the server tried to bind?

kfstorm replied:
done.

rkooo567 reviewed (approved):
A couple last comment!

rkooo567 commented:
Also, can you show me how the error looks like when it is crashed?

kfstorm commented:
@rkooo567 This is the output of raylet.err. (Note that the first line is not in raylet.out.)

```
E0219 10:36:53.510327000 4775783936 server_chttp2.cc:40]               {"created":"@1613702213.510219000","description":"No address added out of total 1 resolved","file":"external/com_github_grpc_grpc/src/core/ext/transport/chttp2/server/chttp2_server.cc","file_line":394,"referenced_errors":[{"created":"@1613702213.510215000","description":"Failed to add any wildcard listeners","file":"external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_server_posix.cc","file_line":341,"referenced_errors":[{"created":"@1613702213.510172000","description":"Unable to configure socket","fd":53,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_server_utils_posix_common.cc","file_line":208,"referenced_errors":[{"created":"@1613702213.510153000","description":"Address already in use","errno":48,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_server_utils_posix_common.cc","file_line":181,"os_error":"Address already in use","syscall":"bind"}]},{"created":"@1613702213.510215000","description":"Unable to configure socket","fd":53,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_server_utils_posix_common.cc","file_line":208,"referenced_errors":[{"created":"@1613702213.510207000","description":"Address already in use","errno":48,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_server_utils_posix_common.cc","file_line":181,"os_error":"Address already in use","syscall":"bind"}]}]}]}
[2021-02-19 10:36:53,510 C 8357 2572821] grpc_server.cc:60:  Check failed: server_ Failed to start the grpc server.
[2021-02-19 10:36:53,510 E 8357 2572821] logging.cc:435: *** Aborted at 1613702213 (unix time) try "date -d @1613702213" if you are using GNU date ***
[2021-02-19 10:36:53,510 E 8357 2572821] logging.cc:435: PC: @                0x0 (unknown)
[2021-02-19 10:36:53,510 E 8357 2572821] logging.cc:435: *** SIGABRT (@0x7fff2052b462) received by PID 8357 (TID 0x11ca8ae00) stack trace: ***
[2021-02-19 10:36:53,510 E 8357 2572821] logging.cc:435:     @     0x7fff2059dd7d _sigtramp
[2021-02-19 10:36:53,516 E 8357 2572821] logging.cc:435:     @        0x10cb71f00 absl::lts_2019_08_08::time_internal::cctz::ZoneInfoSource
[2021-02-19 10:36:53,516 E 8357 2572821] logging.cc:435:     @     0x7fff204ac720 abort
[2021-02-19 10:36:53,516 E 8357 2572821] logging.cc:435:     @        0x10c45295c ray::SpdLogMessage::Flush()
[2021-02-19 10:36:53,517 E 8357 2572821] logging.cc:435:     @        0x10c422c09 ray::RayLog::~RayLog()
[2021-02-19 10:36:53,518 E 8357 2572821] logging.cc:435:     @        0x10c3f3892 ray::rpc::GrpcServer::Run()
[2021-02-19 10:36:53,518 E 8357 2572821] logging.cc:435:     @        0x10bfe71af ray::raylet::NodeManager::NodeManager()
[2021-02-19 10:36:53,518 E 8357 2572821] logging.cc:435:     @        0x10c02b625 ray::raylet::Raylet::Raylet()
[2021-02-19 10:36:53,518 E 8357 2572821] logging.cc:435:     @        0x10c02bbaf ray::raylet::Raylet::Raylet()
[2021-02-19 10:36:53,519 E 8357 2572821] logging.cc:435:     @        0x10bfb71c3 main::$_0::operator()()
[2021-02-19 10:36:53,519 E 8357 2572821] logging.cc:435:     @        0x10bfb5629 std::__1::__function::__func<>::operator()()
[2021-02-19 10:36:53,519 E 8357 2572821] logging.cc:435:     @        0x10c1a11f2 std::__1::__function::__func<>::operator()()
[2021-02-19 10:36:53,520 E 8357 2572821] logging.cc:435:     @        0x10c17a76d _ZZN3ray3rpc12GcsRpcClient17GetInternalConfigERKNS0_24GetInternalConfigRequestERKNSt3__18functionIFvRKNS_6StatusERKNS0_22GetInternalConfigReplyEEEEENKUlS9_SC_E_clES9_SC_
[2021-02-19 10:36:53,520 E 8357 2572821] logging.cc:435:     @        0x10c17a08c ray::rpc::ClientCallImpl<>::OnReplyReceived()
[2021-02-19 10:36:53,521 E 8357 2572821] logging.cc:435:     @        0x10c1bf352 _ZN5boost4asio6detail18completion_handlerIZN3ray3rpc17ClientCallManager29PollEventsFromCompletionQueueEiEUlvE_E11do_completeEPvPNS1_19scheduler_operationERKNS_6system10error_codeEm
[2021-02-19 10:36:53,521 E 8357 2572821] logging.cc:435:     @        0x10c8bafd5 boost::asio::detail::scheduler::do_run_one()
[2021-02-19 10:36:53,521 E 8357 2572821] logging.cc:435:     @        0x10c8ae951 boost::asio::detail::scheduler::run()
[2021-02-19 10:36:53,521 E 8357 2572821] logging.cc:435:     @        0x10c8ae7ec boost::asio::io_context::run()
[2021-02-19 10:36:53,522 E 8357 2572821] logging.cc:435:     @        0x10bfb26b7 main
[2021-02-19 10:36:53,522 E 8357 2572821] logging.cc:435:     @     0x7fff20574621 start
[2021-02-19 10:36:53,522 E 8357 2572821] logging.cc:435:     @               0x19 (unknown)
```

rkooo567 commented:
LGTM!
